src/main/kotlin/com/manimdsl/frontend/SemanticAnalysis.kt

```diff
@@ -308,4 +369,33 @@ class SemanticAnalysis {
             }
         }
     }
+
+    fun allExpressionsAreSameTypeCheck(
+        expected: Type,
+        expressions: List<ExpressionNode>,
+        currentSymbolTable: SymbolTableVisitor,
+        ctx: ParserRuleContext
+    ) {
+        if (expressions.any { inferType(currentSymbolTable, it) != expected }) {
+            inconsistentTypeError(expected, ctx)
+        }
+    }
+
+    fun datastructureConstructorCheck(
+        dataStructureType: DataStructureType,
+        initialValue: List<ExpressionNode>,
+        argumentTypes: List<Type>,
+        ctx: ManimParser.DataStructureContructorContext
+    ) {
+        val constructor = dataStructureType.getConstructor()
+        if (initialValue.isEmpty() && argumentTypes.size < constructor.minRequiredArgsWithoutInitialValue) {
+            missingConstructorArgumentsError(
+                dataStructureType,
+                constructor.minRequiredArgsWithoutInitialValue,
+                argumentTypes.size,
+                ctx
+            )
+        }
+        incompatibleArgumentTypesCheck(dataStructureType, argumentTypes, constructor, ctx)
```

Comment:
Instead of doing this semantic analysis call inside another semantic analysis function, can we move it out into `visitDataStructureConstructor`?

Reply:
Semantically this function checks the whole constructor so I think it is okay